barretenberg/cpp/pil/vm2/alu.pil

```diff
@@ -123,6 +138,83 @@ sel_op_add * (1 - sel_op_add) = 0;
 #[ALU_ADD]
 sel_op_add * (1 - sel_tag_err) * (ia + ib - ic - cf * (max_value + 1)) = 0;
 
+// MUL
+
+sel_op_mul * (1 - sel_op_mul) = 0;
+
+pol commit c_hi;
+
+// MUL - non u128
+
+// TODO(MW): cf required? could be sel_op_mul * (1 - sel_tag_err) * (ia * ib - ic - cf * (max_value + 1) * c_hi) = 0;
+#[ALU_MUL_NON_U128]
+sel_op_mul * IS_NOT_U128 * (1 - sel_tag_err)
+    * (
+        ia * ib
+        - ic
+        - (max_value + 1) * c_hi
+    ) = 0;
+
+// MUL - u128
+
+pol commit sel_mul_u128;
+// sel_op_mul & sel_is_u128:
+sel_mul_u128 - sel_is_u128 * sel_op_mul = 0;
+
+// Taken from vm1:
+// We express a, b in 64-bit slices: a = a_l + a_h * 2^64
+//                                   b = b_l + b_h * 2^64
+// => a * b = a_l * b_l + (a_h * b_l + a_l * b_h) * 2^64 + (a_h * b_h) * 2^128 = c_hi_full * 2^128 + c
+// => the 'top bits' are given by (c_hi_full - (a_h * b_h)) * 2^128
+// We can show for a 64 bit c_hi = c_hi_full - (a_h * b_h) % 2^64 that:
+// a_l * b_l + (a_h * b_l + a_l * b_h) * 2^64 = c_hi * 2^128 + c
+// Equivalently (cf = 0 if a_h & b_h = 0):
+// a * b_l + a_l * b_h * 2^64 = (cf * 2^64 + c_hi) * 2^128 + c
+// => no need for a_h in final relation
+
+pol commit a_lo;
+pol commit a_hi;
+pol commit b_lo;
+pol commit b_hi;
+pol TWO_POW_64 = 2 ** 64;
+
+#[A_MUL_DECOMPOSITION]
+sel_mul_u128 * (ia - (a_lo + TWO_POW_64 * a_hi)) = 0;
+#[B_MUL_DECOMPOSITION]
+sel_mul_u128 * (ib - (b_lo + TWO_POW_64 * b_hi)) = 0;
+
+#[ALU_MUL_U128]
+sel_mul_u128 * (1 - sel_tag_err)
+    * (
+        ia * b_lo + a_lo * b_hi * TWO_POW_64            // a * b without the hi bits
+        - ic                                            // c_lo
+        - (max_value + 1) * (cf * TWO_POW_64 + c_hi)    // c_hi * 2^128 + (cf ? 2^192 : 0)
+    ) = 0;
+
+// TODO: Once lookups support expression in tuple, we can inline constant_64 into the lookup.
+// Note: only used for MUL, so gated by sel_op_mul
+pol commit constant_64;
+sel_op_mul * (64 - constant_64) = 0;
+
+#[RANGE_CHECK_MUL_U128_A_LO]
+sel_mul_u128 { a_lo, constant_64 } in range_check.sel { range_check.value, range_check.rng_chk_bits };
+
+#[RANGE_CHECK_MUL_U128_A_HI]
+sel_mul_u128 { a_hi, constant_64 } in range_check.sel { range_check.value, range_check.rng_chk_bits };
+
+#[RANGE_CHECK_MUL_U128_B_LO]
+sel_mul_u128 { b_lo, constant_64 } in range_check.sel { range_check.value, range_check.rng_chk_bits };
+
+#[RANGE_CHECK_MUL_U128_B_HI]
+sel_mul_u128 { b_hi, constant_64 } in range_check.sel { range_check.value, range_check.rng_chk_bits };
+
+// TODO(MW): range check c_hi for non 128 cases?
+// Each c_hi should pass because it is either from u1-u64 inputs, represent a 64 bit limb of a u128 number, or
+// = 0 for a field, but do we need to check it?
+// No need to range_check c_lo = ic because the memory write will ensure ic <= max_value.
+#[RANGE_CHECK_MUL_U128_C_HI]
+sel_op_mul { c_hi, constant_64 } in range_check.sel { range_check.value, range_check.rng_chk_bits };
```

Comment:
The range check on `c_hi` is definitely required for `u128`s where we could overflow the field, but I'm not sure whether we need it for other tags?
e.g. say we have `u64`s `a` and `b`, we lookup the `mem_tag` info and know that their product cannot overflow the field. If we try and mutate `c` by giving a larger `c_hi`, the mul relation will fail because `max_value` ensures that we 'mod' the overflowed `c` by the correct value. We cannot give an 'unmodded' `c` because the memory write will see its above `u64` and fail.